### ISSUE TITLE
FSDP mixed precision: flag to keep gradients in reduced precision

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#82513]
### 🚀 The feature, motivation and pitch

FSDP mixed precision currently casts gradients back to the parameter full precision type at end of FSDP backwards, this is for the purpose of running optimizer step in full precision.

Although, recent research such as https://arxiv.org/abs/2010.06192 has shown that training can be effective if even optimizer runs in the reduced precision, and only weight updates occur in high precision. To enable this use case, we'd need FSDP to avoid casting the gradients back to the reduced precision. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @ezyang

**Comment by zhaojuanmao** [comment]
for pure bfloat16 training in FSDP, I thought FSDP does not need to do anything, the users just pass a model with bfloat16 to fsdp, fsdp will automatically compute and communicate parameters/gradients in bfloat16, also pass grads with bfloat16 back to optimizers?

**Comment by lessw2020** [comment]
@zhaojuanmao  - what we want to do with this feature is to continue to have weights in FP32 while running FSDP mixed precision, *but* have the local gradients computed with BF16, rather than current upcast to FP32. 

So it's not really for the pure bfloat case.  (as you note above, can just move everything to bfloat before passing to FSDP for that).  
Rather, this is to let the user lower the precision for gradients during mixed precision from FP32 to BF16 and get the memory savings from it.
Based on the work with the BF16 optimizer and the paper, this should result in some memory and perf gains with minimal/no accuracy change.

**Comment by awgu** [comment]
https://github.com/pytorch/pytorch/pull/85223
